File: 1232.check-if-it-is-a-straight-line.cpp
/*
 * @lc app=leetcode id=1232 lang=cpp
 *
 * [1232] Check If It Is a Straight Line
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  bool checkStraightLine(vector<vector<int>>& coordinates) {
    int start_x = coordinates[0][0];
    int start_y = coordinates[0][1];
    int64_t diff_x0 = coordinates[1][0] - start_x;
    int64_t diff_y0 = coordinates[1][1] - start_y;
    int size = coordinates.size();
    for (int i = 2; i < size; i++) {
      int64_t diff_x = coordinates[i][0] - start_x;
      int64_t diff_y = coordinates[i][1] - start_y;
      if (diff_y0 * diff_x != diff_x0 * diff_y) {
        return false;
      }
    }
    return true;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<vector<int>> coordinates;
  coordinates = {{1, 2}, {2, 3}, {3, 4}, {4, 5}, {5, 6}, {6, 7}};
  ASSERT_EQ(s.checkStraightLine(coordinates), true);
  coordinates = {{1, 1}, {2, 2}, {3, 4}, {4, 5}, {5, 6}, {7, 7}};
  ASSERT_EQ(s.checkStraightLine(coordinates), false);
  return 0;
}

File: 1232.check-if-it-is-a-straight-line.rs
/*
 * @lc app=leetcode id=1232 lang=rust
 *
 * [1232] Check If It Is a Straight Line
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn check_straight_line(coordinates: Vec<Vec<i32>>) -> bool {
        let dy = coordinates[1][1] - coordinates[0][1];
        let dx = coordinates[1][0] - coordinates[0][0];
        for i in 2..coordinates.len() {
            let p = &coordinates[i];
            if dx * (p[1] - coordinates[0][1]) != dy * (p[0] - coordinates[0][0]) {
                return false;
            }
        }
        true
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::check_straight_line(vec![
        vec![1, 2],
        vec![2, 3],
        vec![3, 4],
        vec![4, 5],
        vec![5, 6],
        vec![6, 7]
    ]));
}

#[test]
fn test2() {
    assert!(!Solution::check_straight_line(vec![
        vec![1, 1],
        vec![2, 2],
        vec![3, 4],
        vec![4, 5],
        vec![5, 6],
        vec![7, 7]
    ]));
}
